### PR TITLE
Cleanup glue

### PR DESCRIPTION
**Pull request by jysohn23**
* Make sure that MNLI acc metrics are named differently for match vs mismatch
* Allow writing to `cache_dir` from Glue dataset in case where datasets live in readOnly filesystem
* Flush TB writer at end to not miss any metrics

### CODE REVIEW

File: examples/text-classification/run_glue.py [hunk 1]
```diff
@@ -134,24 +134,37 @@ def main():
...
+        if training_args.do_eval
+        else None
+    )
+    test_dataset = (
+        GlueDataset(data_args, tokenizer=tokenizer, mode="test", cache_dir=model_args.cache_dir)
+        if training_args.do_predict
+        else None
+    )
+
+    def build_compute_metrics_fn(task_name: str) -> Dict:
```
Comment (julien-c) [inline examples/text-classification/run_glue.py:151]: typing is incorrect here

Comment (julien-c) [reply examples/text-classification/run_glue.py:151]: Should be `-> Callable[EvalPrediction, Dict]`

File: examples/text-classification/run_glue.py [hunk 2]
```diff
@@ -134,24 +134,37 @@ def main():
...
+        return compute_metrics_fn
 
     # Initialize our Trainer
     trainer = Trainer(
         model=model,
         args=training_args,
         train_dataset=train_dataset,
         eval_dataset=eval_dataset,
-        compute_metrics=compute_metrics,
+        compute_metrics=build_compute_metrics_fn(data_args.task_name),
```
Comment (julien-c) [inline examples/text-classification/run_glue.py:167]: why did you need this? I'm fine with the other changes.

Comment (jysohn23) [reply examples/text-classification/run_glue.py:167]: `task_name = mnli` in line 157 which is then called by: https://github.com/huggingface/transformers/blob/master/src/transformers/trainer.py#L794 which is this function: https://github.com/huggingface/transformers/blob/master/src/transformers/data/metrics/__init__.py#L65-L68 but with `task_name` as `mnli` both whether its `mnli` or `mnli-mm` dataset. This was one way to address it.

File: examples/text-classification/run_glue.py (hunk 1 above)
Comment (jysohn23) [reply examples/text-classification/run_glue.py]: Good point. Done.

**Review by LysandreJik — APPROVED** [review]
LGTM, thanks @jysohn23

File: examples/text-classification/run_glue.py (hunk 2 above)
Comment (julien-c) [reply examples/text-classification/run_glue.py:167]: Yes, but `glue_compute_metrics` handles the case "mnli-mm" correctly right now, no?

Comment (jysohn23) [reply examples/text-classification/run_glue.py:167]: Whether it's `mnli` or `mnli-mm` they both take this code path: https://github.com/huggingface/transformers/blob/master/src/transformers/data/metrics/__init__.py#L67-L68

**Comment by jysohn23** [comment]
@julien-c bump..!

File: src/transformers/data/datasets/glue.py [hunk 3]
```diff
@@ -70,6 +70,7 @@ def __init__(
         tokenizer: PreTrainedTokenizer,
         limit_length: Optional[int] = None,
         mode: Union[str, Split] = Split.train,
+        cache_dir: str = None,
```
Comment (julien-c) [inline src/transformers/data/datasets/glue.py:73]:
```suggestion
        cache_dir: Optional[str] = None,
```
